Trivial consistency check

Nemo157 opened the pull request:
part of #766

Nemo157 commented on src/index/crates.rs:
```diff
@@ -0,0 +1,55 @@
+
+        tree.walk(git2::TreeWalkMode::PreOrder, |_, entry| {
+            result = (|| {
+                if let Some(blob) = entry.to_object(&self.repo)?.as_blob() {
```
This is getting the content, not the name.

jyn514 replied:
I don't know the git2 API well enough - can't you pass the name of the file to crates_index so you don't need a temporary file? Or does crates_index assume that the file is writable or something?

Nemo157 replied:
The file doesn't actually exist anywhere, it is stored as part of git's compressed archives. That's why we read out the content and write it into a temporary named file that we can pass to `crates_index` to read back in and parse. I planned to take a look at whether it's possible to expose the [`crates_index::Crate::from_slice`](https://docs.rs/crates-index/0.15.0/src/crates_index/lib.rs.html#376-396) constructor so that we could just call that with the blob content directly (assuming that that's easier than changing `crates_index` to support bare git repos itself directly).

jyn514 replied:
I opened https://github.com/frewsxcv/rust-crates-index/issues/38, it _looks_ like they just have to add `pub` in front of the method.

jyn514 replied:
They just published a new version with the change :) https://crates.io/crates/crates-index/0.15.1

Nemo157 commented on src/utils/consistency/mod.rs:
```diff
@@ -0,0 +1,58 @@
+    conn: &postgres::Connection,
+    dry_run: bool,
+) -> Result<(), failure::Error> {
+    if !dry_run {
+        failure::bail!("TODO: only a --dry-run synchronization is supported currently");
+    }
```
A quick check against the prod-dump from april (with my index force reset back to then) shows 4159 releases of 1477 crates in the index that are not in the db, and 1482 releases of 1438 crates in the db that are not in the index. There's definitely more work to do before this could be run to fix stuff (mostly around improving queueing behaviour), but I think we can get to a point where it's safe enough to have a non dry-run version of this, and manually fixing the issues seems like it's too large.

jyn514 commented:
You had a couple `cargo-timing.html` files that snuck in, please remove those.

Nemo157 commented:
Removed the `cargo-timing.html`, updated `crates-index` and rebased over master (conflicted with #917 since that had cherry-picked changes from here).

jyn514 commented on src/lib.rs:
```diff
@@ -3,6 +3,7 @@
 #![allow(clippy::cognitive_complexity)]
 
 pub use self::build_queue::BuildQueue;
+
```
```suggestion
```

jyn514 commented on src/index/crates.rs:
```diff
@@ -0,0 +1,47 @@
+            })()
+            .with_context(|_| {
+                format!(
+                    "Loading crate details from '{}'",
```
```suggestion
                    "loading crate details from '{}'",
```

jyn514 commented on src/index/crates.rs:
```diff
@@ -0,0 +1,47 @@
+                    if let Ok(krate) = Crate::from_slice(blob.content()) {
+                        f(krate);
+                    } else {
+                        log::warn!("Not a crate '{}'", entry.name().unwrap());
```
Errors should be lowercase: https://rust-lang.github.io/api-guidelines/interoperability.html#error-types-are-meaningful-and-well-behaved-c-good-err
```suggestion
                        log::warn!("not a crate '{}'", entry.name().unwrap());
```
